Convert to a plugin

The right way to work with hapi is to build your application using plugins. Plugins provide a way to organize your code into logical components and then put them together in different combinations or deploy them in different configurations. While some plugins are published as general purpose utilities (e.g. adding authentication), you should think of plugins as a way to break your code into pieces.

Now that we have a basic server with one endpoint, we want to move the creation of the endpoint to a [plugin](http://hapijs.com/tutorials/plugins), and then register that plugin with the server. Creating a new file `lib/version.js` and move the `/version` endpoint there, wrapped in the plugin `register()` function.

Then change our current `lib/index.js` to require the new file and register the version plugin with our server.

Remember to follow the style guide, and ask any questions right here in the comments. If you are not sure how to get your fork back in sync with the current updated code, use the [git guide](https://github.com/hueniverse/hueniversity/blob/master/guides/git.md).

Assignment due: 3/24

I have now this.

index.js : 

~~~~

// Load modules

var Hapi = require('hapi');
var Hoek = require('hoek');
var Package = require('../package.json');


// Declare internals

var internals = {};


internals.init = function () {
  
  var server = new Hapi.Server();
  server.connection({ port: 8000 });

  server.register({register: require('version')}, function (err) {
    
    if (err) {
        console.error('Failed to load plugin:', err);
    }
   });
  
  server.start(function (err) {

  Hoek.assert(!err, err);
  console.log('Server started at: ' + server.info.uri);
  });
};

internals.init();

~~~

version.js

~~~

exports.register = function (server, options, next) {
  
  server.route({
    
    method: 'GET',
    path: '/version',
        config: {
          description: 'Returns the version of the server',
          handler: function (request, reply) {

                return reply({ version: Package.version });
            }
        }
    });
    
  next();
};

exports.register.attributes = {
  name: "version",
  version: "0.0.1"
};

~~~ 

but still I see this error message : 

Error: Cannot find module 'version'

Roelof

--- a/lib/index.js
+++ b/lib/index.js
@@ -2,35 +2,24 @@
 
 var Hapi = require('hapi');
 var Hoek = require('hoek');
-var Package = require('../package.json');
-
+var Version = require('./version');
 
 // Declare internals
 
 var internals = {};
 
-
 internals.init = function () {
 
     var server = new Hapi.Server();
     server.connection({ port: 8000 });
-
-    server.route({
-        method: 'GET',
-        path: '/version',
-        config: {
-            description: 'Returns the version of the server',
-            handler: function (request, reply) {
-
-                return reply({ version: Package.version });
-            }
-        }
-    });
-
-    server.start(function (err) {
-
+    server.register(Version, function (err) {
+      
         Hoek.assert(!err, err);
-        console.log('Server started at: ' + server.info.uri);
+        server.start(function (err) {
+
+            Hoek.assert(!err, err);
+            console.log('Server started at: ' + server.info.uri);
+        });
     });
 };
 

--- a/lib/version.js
+++ b/lib/version.js
@@ -1,0 +1,29 @@
+// Declare internals
+var internals = {
+  response: {
+    version: require('../package.json').version
+  }
+};
+
+// Register plugin
+exports.register = function (server, options, next) {
+
+    server.route({
+        method: 'GET',
+        path: '/version',
+        config: {
+            description: 'Returns the version of the server',
+            handler: function (request, reply) {
+
+                return reply(internals.response);
+            }
+        }
+    });
+
+    return next();
+};
+
+// Plugin metadata
+exports.register.attributes = {
+    name: 'version'
+};

